core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -187,47 +201,78 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
         }
         .flatMap { entry =>
           try {
-            Some(replay(entry, new ReplayListenerBus()))
+            Some((getModificationTime(entry).get, entry))
           } catch {
             case e: Exception =>
               logError(s"Failed to load application log data from $entry.", e)
               None
           }
         }
-        .sortWith(compareAppInfo)
+        .sortWith(_._1 >= _._1)
+        .map(_._2)
+
+      val itor = logInfos.iterator
+      while (itor.hasNext) {
+        val batchSize = 20
+        val batch = itor.take(batchSize)
+        replayExecutor.submit(new Runnable {
+          override def run(): Unit = replay(batch)
+        })
+        for (i <- 1 to batchSize) {
+          if(itor.hasNext) {
+            itor.next()
+          }
+        }
+      }
 
       lastModifiedTime = newLastModifiedTime
+    } catch {
+      case e: Exception => logError("Exception in checking for event log updates", e)
+    }
+  }
 
-      // When there are new logs, merge the new list with the existing one, maintaining
-      // the expected ordering (descending end time). Maintaining the order is important
-      // to avoid having to sort the list every time there is a request for the log list.
-      if (!logInfos.isEmpty) {
-        val newApps = new mutable.LinkedHashMap[String, FsApplicationHistoryInfo]()
-        def addIfAbsent(info: FsApplicationHistoryInfo) = {
-          if (!newApps.contains(info.id) ||
-              newApps(info.id).logPath.endsWith(EventLoggingListener.IN_PROGRESS) &&
-              !info.logPath.endsWith(EventLoggingListener.IN_PROGRESS)) {
-            newApps += (info.id -> info)
-          }
-        }
+  /**
+   * Fetch and Parse the log files
+   */
+  private def replay(logs: Iterator[FileStatus]) {
```

Comment:
Also, just for clarity, I'd call this method `parseApplicationInfos` or something, since it's doing more than just replaying the logs.